### ISSUE TITLE
Node serving an external client may crash [with a proposed bugfix]

### ISSUE THREAD

**Issue by songweijia** [Derecho-Project/derecho#214]
A node serving an external client may crash with some rare cases. When an external client disconnects from a node after calling a remote void function with `p2p_send`, the serving node might haven't finish handing that request. After a while, when the node finished handling and prepare to send a simulated reply in `RPCManager::p2p_request_worker()`, the node calls `connections->get_sendbuffer_ptr()` for a reply buffer. However, `get_sendbuffer_ptr()` will return a `nullptr` because the external client has already disconnected at this moment. Without checking the return value of `get_sendbuffer_ptr()`, a segmentation fault crashes the serving node. We should check if the buffer pointer is valid and only fill and send the simulated reply when it is valid.

Please see the proposed fix below.
```
diff --git a/src/core/rpc_manager.cpp b/src/core/rpc_manager.cpp
index 2ba249e9..76d4c99e 100644
--- a/src/core/rpc_manager.cpp
+++ b/src/core/rpc_manager.cpp
@@ -432,8 +432,10 @@ void RPCManager::p2p_request_worker() {
         } else {
             // hack for now to "simulate" a reply for p2p_sends to functions that do not generate a reply
             char* buf = connections->get_sendbuffer_ptr(request.sender_id, sst::REQUEST_TYPE::P2P_REPLY);
-            buf[0] = 0;
-            connections->send(request.sender_id);
+            if (buf != nullptr) {
+                buf[0] = 0;
+                connections->send(request.sender_id);
+            }
         }
     }
 }
```

**Comment by etremel** [comment]
This fix looks good to me. In general we should be checking the return value of `get_sendbuffer_ptr()` before using it, and not checking it in that case in RPCManager was an oversight.

**Comment by songweijia** [comment]
Fix merged to master at commit 0566953.
